### PR TITLE
feat: add pause barrier and resume barrier

### PR DESCRIPTION
**Pull request by KeXiangWang**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

Add pause barrier and resume barrier.
Once received pause barrier, the source executor will stop to read from source and only receive new barriers.
Scale service in Meta and two interfaces are added to do testing and controling.

## Checklist

- [x] I have written necessary docs and comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Refer to a related PR or issue link (optional)
https://github.com/singularity-data/risingwave/issues/3292

### CODE REVIEW

**Comment by skyzh** [comment]
Any e2e test / manual testing / bench result?

**Comment by skyzh** [comment]
May also add pause / resume to risectl so it’s easier to do testing.

**Comment by KeXiangWang** [comment]
> Any e2e test / manual testing / bench result?

manually tested.

**Comment by KeXiangWang** [comment]
> May also add pause / resume to risectl so it’s easier to do testing.

It's a part of config change. Since we havent implemented the whole config change process, I added these two interfaces. They should be removed when config change is ready.

**Comment by BugenZhao** [comment]
Will review it later. Please hold on this PR for a while. :)

**Comment by skyzh** [comment]
I think of this PR as a draft, as it doesn't implement full functionality. I'll need more time to revisit source executor on possible racing condition next week. May ping @BugenZhao for a preliminary review and get it merged. Some design concerns:

* Should we persist pause / resume state to meta store so that after cluster failure we will still pause everything?
* Should newly added actors be paused is a pause barrier is sent beforehand?
* If source is paused, INSERT statement will stuck. Is it acceptable for users?
* Source executor now has two streams communicating through channels, which may have some racing conditions. Can we ensure that there's no more message after pause barrier is produced from source executor in current implementation?

There's no need to solve all of the issues mentioned above, but I think they should be decided throughout the future implementation.

**Comment by skyzh** [comment]
Also pause doesn't affect chain executor. Do you expect chain to produce data after pause barrier is sent?

**Comment by tabVersion** [comment]
> If source is paused, INSERT statement will stuck. Is it acceptable for users?

I think we can merely pause connector source and leave table source alive. This can help user do INSERT during scaling.

**Comment by fuyufjh** [comment]
* Should we persist pause / resume state to meta store so that after cluster failure we will still pause everything?

Not necessary, because it's meaningless to recover to a "paused" state. 

BTW, the 'pause' barrier should trigger a checkpoint because other Compute Nodes will rely on this to recover their state.

* Should newly added actors be paused if a pause barrier is sent beforehand?

Yes, the newly added actors should be in a paused status. 

In my original design, MV-on-MV also works in the same way. But since we implemented MV-on-MV first without this, the newly-added actors there were not paused first.

* If source is paused, INSERT statement will stuck. Is it acceptable for users?

Yes, but the pause-resume process should be very fast - No IO waiting happens during this. The most time-consuming part is the newly-created actors recover state from checkpoint, but the waiting will not block this process.

* Source executor now has two streams communicating through channels, which may have some racing conditions. Can we ensure that there's no more message after pause barrier is produced from source executor in current implementation?

This is guaranteed by design.

But I also worry about Rust compiler will complain a lot about this... Didn't think deeper...

File: src/stream/src/executor/source.rs [hunk 1]
```diff
@@ -147,45 +154,78 @@ impl SourceReader {
...
-                        break 'outer;
+            } else {
+                // paused
+                let msg = abort_notifier.recv().await.unwrap();
+                if let AbortMessage::Resume(_) = msg {
+                    if reader_tmp.is_some() {
+                        let mut reader_guard = stream_reader.lock().await;
+                        if reader_guard.is_none() {
+                            *reader_guard = reader_tmp.take();
+                        } else {
```
Comment (tabVersion) [inline src/stream/src/executor/source.rs:222]: in which case will the `else` be triggered? 

rest parts in source LGTM

**Comment by KeXiangWang** [comment]
> Should we persist pause / resume state to meta store so that after cluster failure we will still pause everything?

Haven't consider too much about failure when config change. Intuitively, if we recover directly based on new configuration, there's no need keep the `paused` state.

Checkpointing will be triggered when the pause/resume barrier flow through the graph, just like other barriers.

> Should newly added actors be paused if a pause barrier is sent beforehand?

They're naturally paused, since we do not scale the source executors now, there won't be data on the fly when the new actors are created.

> If source is paused, INSERT statement will stuck. Is it acceptable for users?

Like Eric said, I suppose it's acceptable.

> Source executor now has two streams communicating through channels, which may have some racing conditions. Can we ensure that there's no more message after pause barrier is produced from source executor in current implementation?

It's ensured that no more chunks will be produced. The barriers can still flow through the graphs as usual.

> Also pause doesn't affect chain executor. Do you expect chain to produce data after pause barrier is sent?

Sorry, I didn't think of it. Will consider this before merging.

**Comment by skyzh** [comment]
> It's ensured that no more chunks will be produced. The barriers can still flow through the graphs as usual

I'm thinking about the ordering of pause barrier and data chunks. Not sure whether the current implementation ensures that barriers always come after chunks, as we prefer barrier in selects.

**Comment by skyzh** [comment]
> They're naturally paused, since we do not scale the source executors now, there won't be data on the fly when the new actors are created, until the resume barriers are sent

Actors can also be added by create materialized view, so I believe you should implement and test this.

File: src/stream/src/executor/mod.rs [hunk 2]
```diff
@@ -177,6 +178,8 @@ pub enum Mutation {
     UpdateOutputs(HashMap<(ActorId, DispatcherId), Vec<ActorInfo>>),
     AddOutput(AddOutput),
     SourceChangeSplit(HashMap<ActorId, ConnectorState>),
+    Pause(()),
+    Resume(()),
```
Comment (BugenZhao) [inline src/stream/src/executor/mod.rs:182]:
```suggestion
    Pause,
    Resume,
```

File: src/stream/src/executor/source.rs [hunk 3]
```diff
@@ -34,6 +34,13 @@ use super::error::StreamExecutorError;
 use super::monitor::StreamingMetrics;
 use super::*;
 
+#[derive(Debug)]
+enum AbortMessage {
+    Pause(()),
+    Resume(()),
+    SourceChange(()),
```
Comment (BugenZhao) [inline src/stream/src/executor/source.rs:41]:
```suggestion
    Pause,
    Resume,
    SourceChange,
```

File: src/stream/src/executor/source.rs [hunk 4]
```diff
@@ -350,30 +390,40 @@ impl<S: StateStore> SourceExecutor<S> {
...
+                                                self.stream_source_splits = target_state;
+                                            }
                                         }
                                     }
+                                    Mutation::Pause(_) => {
+                                        abort_tx.send(AbortMessage::Pause(())).unwrap();
+                                    }
+                                    Mutation::Resume(_) => {
+                                        abort_tx.send(AbortMessage::Resume(())).unwrap();
+                                    }
```
Comment (BugenZhao) [inline src/stream/src/executor/source.rs:425]: Do we really need to make the source reader aware of `Pause` and `Resume`? IMHO, if we do not poll the reader here, then we're already paused.

Comment (tabVersion) [reply src/stream/src/executor/source.rs:425]: yes, but polling source reader happens in a separate thread, we should tell it when to pause and when to resume

**Comment by KeXiangWang** [comment]
> > They're naturally paused, since we do not scale the source executors now, there won't be data on the fly when the new actors are created, until the resume barriers are sent
> 
> Actors can also be added by create materialized source / create source + create MV, so I believe you should implement and test this.

IMHO, we can simply block all these DDL sql when doing config change. Since we already paused the data ingestion, it sounds reasonable we also pause these DDL statements, which also makes the config change easier to implement.

File: src/stream/src/executor/source.rs (hunk 3 above)
Comment (KeXiangWang) [reply src/stream/src/executor/source.rs]: done

File: src/stream/src/executor/mod.rs (hunk 2 above)
Comment (KeXiangWang) [reply src/stream/src/executor/mod.rs:182]: done

**Comment by skyzh** [comment]
> we can simply block all these DDL sql

Looks good in this stage. Can improve later.

Also currently all current design looks to assume that scaling will always succeed. What if scale service crashes and resume barrier is never sent? Better to take this into consideration in the future.

**Comment by KeXiangWang** [comment]
Updated. @skyzh @BugenZhao

File: src/stream/src/executor/mod.rs [hunk 5]
```diff
@@ -179,6 +180,8 @@ pub enum Mutation {
...
```
Comment (BugenZhao) [inline src/stream/src/executor/mod.rs:184]:
```suggestion
    Pause,
    Resume,
```

Comment (KeXiangWang) [reply src/stream/src/executor/mod.rs]: updated

**Comment by BugenZhao** [comment]
> > > They're naturally paused, since we do not scale the source executors now, there won't be data on the fly when the new actors are created, until the resume barriers are sent
> 
> > 
> 
> > Actors can also be added by create materialized source / create source + create MV, so I believe you should implement and test this.
> 
> 
> 
> IMHO, we can simply block all these DDL sql when doing config change. Since we already paused the data ingestion, it sounds reasonable we also pause these DDL statements, which also makes the config change easier to implement. 

I think these two barriers will not be issued manually after we fully implement scaling. At that time we'll ensure that the pause/resume and the scaling barrier are executing atomically so there will be no problem. Personally I think these changes are acceptable.

**Comment by KeXiangWang** [comment]
> > Also pause doesn't affect chain executor. Do you expect chain to produce data after pause barrier is sent?
> 
> Has this been resolved? Seems critical to the correctness.

Working on that.

**Review by skyzh — APPROVED** [review]
The implementation will need refinement in the future, but currently it looks good enough.

* What if client disconnects when calling the create_mv RPC endpoint? The lock is immediately dropped and bad things can happen.
* What if we're having more DDL types and RPCs? Developers might forget to add the lock, and there will be potential issues.

I think it would be good to have some control in barrier manager. It should schedule the barriers to send. e.g. in `run_command`, if there's ongoing DDL commands running, we should wait. Having a lock outside of barrier manager is okay for now. Maybe we can make it better in later PRs.

I'd also like to wait for @BugenZhao 's comments.

**Comment by KeXiangWang** [comment]
> The implementation will need refinement in the future, but currently it looks good enough.
> 
> * What if client disconnects when calling the create_mv RPC endpoint? The lock is immediately dropped and bad things can happen.
> * What if we're having more DDL types and RPCs? Developers might forget to add the lock, and there will be potential issues.
> 
> I think it would be good to have some control in barrier manager. It should schedule the barriers to send. e.g. in `run_command`, if there's ongoing DDL commands running, we should wait. Having a lock outside of barrier manager is okay for now. Maybe we can make it better in later PRs.
> 
> I'd also like to wait for @BugenZhao 's comments.

Thanks for the comments!

**Pull request merged**